### PR TITLE
`Component` proper setter/getter, constructor may set its `author`, normalize optional `version`

### LINKED ISSUE #246: Include `component.author` node in generated SBOM
I noticed that the `author` node from the `package.json` file isn't included in the generated SBOM. I would expect this value to be copied to the SBOM `metadata.component.author` node.

### LINKED ISSUE #248: JSON output does not guarantee `component.version` value
I was using the JSON output of this tool as an input for another tool.  I implemented a validation function in python for my other tool in order to ensure its correctness before processing the file, and can be seen as follows:

```
import json
import jsonschema
import pkg_resources

def validate_bom(filename):
    bom = json.load(open(filename, "r"))
    spec_version = bom.get("specVersion", None)
    schema_path = pkg_resources.resource_filename('cyclonedx', f'schema/bom-{spec_version}.schema.json')
    bom_schema = json.load(open(schema_path, "r"))
    try:
        jsonschema.validate(instance=bom, schema=bom_schema)
    except jsonschema.exceptions.ValidationError as err:
        raise err
    return True
```

In this case I was using I created a BoM using schema v1.3 and noticed that there was a `version` value missing from the component within the optional `metadata` property, and was throwing an error as invalid.  

In the schema file [bom-1.3.schema.json](https://cyclonedx.org/schema/bom-1.3.schema.json), the optional metadata section, allows for components to be listed and references the component in the `#/definitions/component` section, where a version is required.  In this case, there was no version associated with the component found in the metadata secion.

The output is therefore non-compliant to the v1.3 schema standard, and the output _should_ be verified to ensure that it is compliant before outputting to the user.

### PR DESCRIPTION
**Pull request by jkowalleck**
- detect component.author - fixes #246
- normalize empty component.version - fixes #248
- add JSDoc - #220
- use private properties - #233
- `Component` uses setters, so that type-checks are done as expected.
- bumped dependencies (#235 & #250)
- pinned dev dependencies to a specific version in npm manifest
- eslint uses babel parser to detect stage4 features - for #233
- added additiona CI/CT tests

### CODE REVIEW

**Comment by shiraze** [comment]
I think code should follow existing pattern, so:
```
this._author = (pkg.author?.name) ? pkg.author.name : undefined
```

And the expectation is that `package.json` would contain
```
...
 "author": { "name": "a name" },
...
```

Having said that, though, your code change would also be correct, but wouldn't it be better to use the "private" _author field?

**Comment by jkowalleck** [comment]
re https://github.com/CycloneDX/cyclonedx-node-module/pull/247#issuecomment-1022493733
i thing the current code base is a mess and i dont want to continue like this anymore.
current code has proper setters for everything, and they do proper type checking - which is required to produce the proper XML/JSON results.

WHY on earth would one use the internal properties that are not type safe at all, if there are proper setters with type checks? 

no, in don't think i will use the internal properties. i'll use the existing setters.

but i will use the optional-property style you suggested, @shiraze . thanks for the hint :+1:

**Review by shiraze — CHANGES_REQUESTED** [review]
As discussed in conversation set using optional property style, and we're all good :)

**Comment by jkowalleck** [comment]
na, cannot use shorthand ala `pkg.author?.name` - it is not supported by node12 - which is the lowest supported version. 
see https://github.com/CycloneDX/cyclonedx-node-module/runs/4972957766?check_suite_focus=true

File: model/Component.js [hunk 1]
```diff
@@ -26,31 +26,108 @@ const ExternalReferenceList = require('./ExternalReferenceList')
...
 
+/**
+ * @typedef {"application"|"framework"|"library"|"container"|"operating-system"|"device"|"firmware"|"file"} ComponentType
+ * @typedef {"required"|"optional"|"excluded"} ComponentScope
+ */
 class Component extends CycloneDXObject {
+
+  // required properties
+  /** @type {ComponentType} */
+  #type = 'library'
```
[the remark this answers is not in the document]
Comment (jkowalleck) [reply model/Component.js:38]: @sonatype-lift help

[the remark this answers is not in the document]
Comment (jkowalleck) [reply model/Component.js:38]: @sonatype-lift ignore

[the remark this answers is not in the document]
Comment (jkowalleck) [reply model/Component.js:38]: @sonatype-lift ignore

[the remark this answers is not in the document]
Comment (jkowalleck) [reply model/Component.js:38]: @sonatype-lift ignore

File: model/Component.js [hunk 2]
```diff
@@ -63,189 +140,336 @@ class Component extends CycloneDXObject {
...
-    this._bomRef = this.validateType('bom-ref', value, String)
+    this.#bomRef = this.validateType('bom-ref', value, String)
   }
 
+  /**
+   * @returns {OrganizationalEntity|undefined}
+   */
   get supplier () {
-    return this._supplier
+    return this.#supplier
```
[the remark this answers is not in the document]
Comment (jkowalleck) [reply model/Component.js:196]: @sonatype-lift ignore

File: model/Component.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (jkowalleck) [reply model/Component.js:38]: @sonatype-lift ignore

File: model/Component.js [hunk 3]
```diff
@@ -63,189 +140,336 @@ class Component extends CycloneDXObject {
...
-    return this._bomRef
+    return this.#bomRef
   }
 
+  /**
+   * @param {string|undefined} value
+   */
   set bomRef (value) {
-    this._bomRef = this.validateType('bom-ref', value, String)
+    this.#bomRef = this.validateType('bom-ref', value, String)
```
[the remark this answers is not in the document]
Comment (jkowalleck) [reply model/Component.js:189]: @sonatype-lift ignore

File: model/Component.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (jkowalleck) [reply model/Component.js:37]: @sonatype-lift unignore

[the remark this answers is not in the document]
Comment (jkowalleck) [reply model/Component.js:46]: fixed

File: .github/workflows/nodejs.yml [hunk 4]
```diff
@@ -26,9 +26,11 @@ jobs:
         node-version: 
           # action based on https://github.com/actions/node-versions/releases
           # see also: https://nodejs.org/en/about/releases/
-          - "16"  # active LTS
+          - "17"      # current
```
Comment (brent-spiner) [inline .github/workflows/nodejs.yml:29]: Seeing this just kinda reminded me that odd: 6 months, and even, 30 months. Can you start testing 18 yet or is it still pie in the sky?

**Comment by jkowalleck** [comment]
@brent-spiner made my day

**Pull request merged**